src/runtime_src/core/pcie/windows/shim.cpp

```diff
@@ -1610,3 +1610,387 @@ xclReadTraceData(xclDeviceHandle handle, void* traceBuf, uint32_t traceBufSz,
 
   return size;
 }
+
+/*
+ * Returns  1 if Version2 > Version1
+ * Returns  0 if Version2 = Version1
+ * Returns -1 if Version2 < Version1
+ */
+signed cmpMonVersions(unsigned major1, unsigned minor1, unsigned major2, unsigned minor2) {
+  if (major2 > major1)
+    return 1;
+  else if (major2 < major1)
+    return -1;
+  else if (minor2 > minor1)
+    return 1;
+  else if (minor2 < minor1)
+    return -1;
+  else return 0;
+}
+
+
+// Gets the information about the specified IP from the sysfs debug_ip_table.
+// The IP types are defined in xclbin.h
+uint32_t getDebugIpData(xclDeviceHandle handle, int type, uint64_t *baseAddress, std::string * portNames,
+                                       uint8_t *properties, uint8_t *majorVersions, uint8_t *minorVersions,
+                                       size_t size)
+{
+  auto shim = get_shim_object(handle);
+  size_t sz1 = 0, sectionSz = 0;
+  // Get the size of full debug_ip_layout
+  shim->get_debug_ip_layout(nullptr, sz1, &sectionSz);
+  // Allocate buffer to retrieve debug_ip_layout information from loaded xclbin
+  std::vector<char> buffer(sectionSz);
+  shim->get_debug_ip_layout(buffer.data(), sectionSz, &sz1);
+  debug_ip_layout *map = reinterpret_cast<debug_ip_layout*>(buffer.data());
+  uint32_t count = 0;
+  for( unsigned int i = 0; (i < map->m_count) && (count < size) ; i++ ) {
+    if(map->m_debug_ip_data[i].m_type != type)
+      continue;
+
+    if(baseAddress) baseAddress[count] = map->m_debug_ip_data[i].m_base_address;
+    if(portNames) {
+      // Fill up string with 128 characters (padded with null characters)
+      portNames[count].assign(map->m_debug_ip_data[i].m_name, 128);
+      // Strip away extraneous null characters
+      portNames[count].assign(portNames[count].c_str());
```

Comment:
The m_name entry in debug_ip_data is guaranteed to have a null character at the end of the string.  Why don't you remove this line and replace line 1654 with the following code:
`portNames[count].assign(&(map->m_debug_ip_data[i].m_name));`